keystone/rebaseline_multi.py

```diff
@@ -0,0 +1,173 @@
+	middle = numpy.std(skimage.util.view_as_windows(spectra, 31, 1),axis=1)
+	stds = numpy.concatenate((left, middle, right))
+	mask = numpy.arange(spec_len)[numpy.argsort(stds)[:sample]]
+	#median_std = numpy.std(stds)*3
+	#mask = numpy.where(numpy.array(stds)<median_std)[0]
+	#mask = numpy.concatenate((mask, numpy.arange(-50, 50)))
+	#if 1>0: #len(mask)<30
+		#print 'yes'
+		#mask = numpy.arange(len(spectra))[numpy.argsort(stds)[:500]]
+		#mask=numpy.arange(len(spectra))[0::5]
+
+	#plt.plot(range(len(spectra)), spectra)
+	#plt.plot(numpy.arange(len(spectra))[mask], spectra[mask])
+	#plt.show()
+	return mask
+
+def redchisqg(ydata,ymod,deg=2,sd=None):     
+      """Returns the reduced chi-square error statistic for an arbitrary model,   
+ chisq/nu, where nu is the number of degrees of freedom. If individual   
+ standard deviations (array sd) are supplied, then the chi-square error   
+ statistic is computed as the sum of squared errors divided by the standard   
+ deviations. See http://en.wikipedia.org/wiki/Goodness_of_fit for reference.  
+   
+ ydata,ymod,sd assumed to be Numpy arrays. deg integer.  
+   
+ Usage:  
+ >>> chisq=redchisqg(ydata,ymod,n,sd)  
+ where  
+  ydata : data  
+  ymod : model evaluated at the same x points as ydata  
+  n : number of free parameters in the model  
+  sd : uncertainties in ydata  
+   
+ Rodrigo Nemmen  
+ http://goo.gl/8S1Oo"""  
+      # Chi-square statistic  
+      if sd==None:  
+           chisq=numpy.sum((ydata-ymod)**2)  
+      else:  
+           chisq=numpy.sum( ((ydata-ymod)/sd)**2 )  
+             
+      # Number of degrees of freedom assuming 2 free parameters  
+      nu=ydata.size-1-deg  
+        
+      return chisq/nu
+
+def get_chi(blorder, ydata, xdata, blindex, noise):
+    """
+ Returns the best-fit Legendre polynomial to an input spectra, 
+ along with that model's reduced chi-squared value.     
+   
+ blorder = order of polynomial to fit
+ ydata = spectrum y values
+ xdata = spectrum x values
+ blindex = the indices of the spectra to include in baseline fitting
+ noise = rms noise of spectrum"""
+    opts = lsq(legendreLoss, np.zeros(blorder + 1), args=(ydata[blindex],
+                                                          xdata[blindex],
+                                                        noise), loss='arctan')
+    ymod = legendre.legval(xdata, opts.x)
+    chi = redchisqg(ydata,ymod,deg=blorder+1)
+    return ymod, chi
+
+def robustBaseline_chi(y, baselineIndex, blorder_max=3, noiserms=None):
+    """  
+ Returns a baseline subtracted spectrum, based on the best-fitting polynomial 
+ for a range of polynomial orders.    
+   
+ y = input spectra
+ baselineIndex = indices of spectra to include in baseline fitting
+ blorder_max = largest order polynomial to fit (fit from blorder_max down to order of 1) 
+ noiserms = rms noise of spectrum"""
+    x = np.linspace(-1, 1, len(y))
+    if noiserms is None:
+        noiserms = mad1d((y - np.roll(y, -2))[baselineIndex])
```

Comment:
Things are clearly working and this isn't incredibly critical, but the noiserms should be divided by sqrt(2) since the difference between two Gaussian deviates has a distribution that is sqrt(2) larger than the contributing distributions.

Reply:
Ok I see now that sqrt(2) term is in the keystone.baseline.rebaseline function.  I used the keystone.baseline.robustBaseline as the template for this function.  There the sqrt(2) term isn't included when the noiserms parameter isn't supplied.  Is that by design? Or should the sqrt(2) term be there also?

Reply:
Thanks.  I fixed this too.